zerver/views/message_edit.py

```diff
@@ -91,6 +92,35 @@ def get_message_edit_history(
 PROPAGATE_MODE_VALUES = ["change_later", "change_one", "change_all"]
 
 
+def validate_message_edit_payload(
+    message: Message,
+    stream_id: Optional[int],
+    topic_name: Optional[str],
+    propagate_mode: Optional[str],
+    content: Optional[str],
+) -> None:
+    if topic_name is None and content is None and stream_id is None:
+        raise JsonableError(_("Nothing to change"))
+
+    if not message.is_stream_message() and stream_id is not None:
+        raise JsonableError(_("Private messages cannot be moved to streams."))
+
+    if propagate_mode != "change_one" and topic_name is None and stream_id is None:
+        raise JsonableError(_("Invalid propagate_mode without topic edit"))
+
+    if topic_name is not None:
+        topic_name = topic_name.strip()
+        if topic_name == "":
+            raise JsonableError(_("Topic can't be empty"))
+
+    if stream_id is not None and content is not None:
+        raise JsonableError(_("Cannot change message content while changing stream"))
+
+    # Right now, we prevent users from editing widgets.
+    if content is not None and is_widget_message(message):
+        raise JsonableError(_("Widgets cannot be edited."))
```

Comment:
We might also want to put this function in a `zerver/lib/message.py` type file so that it can be called from `outgoing_webhook.py`, see my top-level comment for details.  @PIG208 FYI.